packages/ckeditor5-ui/src/panel/balloon/balloonpanelview.ts

```diff
@@ -387,6 +404,27 @@ export default class BalloonPanelView extends View {
 		this.listenTo( global.window, 'resize', () => {
 			this.attachTo( options );
 		} );
+
+		// Hide the panel if the target element is no longer visible.
+		if ( targetElement && !this._resizeObserver ) {
+			const checkVisibility = () => {
+				// If the target element is no longer visible, hide the panel.
+				if ( !isVisible( targetElement ) ) {
+					this.unpin();
+				}
+			};
+
+			// Element is being resized to 0x0 after it's parent became hidden,
+			// so we need to check size in order to determine if it's visible or not.
+			this._resizeObserver = new ResizeObserver( targetElement, checkVisibility );
+
+			// Check the visibility of the target element immediately, as the target element might be
+			// already hidden when the panel is being pinned. It's wrapped in `requestAnimationFrame` because the
+			// `_startPinning` method is called during `pin()` and the panel is not yet visible, and not every listener
+			// has been attached yet. This way, we wait until rest of the listeners are attached in caller methods
+			// and then check the visibility.
+			window.requestAnimationFrame( checkVisibility );
```

Comment:
I'm not sure about this approach. Can we listen to some event or observable property to know when it's ready for this check?

Reply:
@niegowski I'm not sure if we can do that without breaking integrations. `.pin()` used to return `void` so I assume that every integration assumes that balloon is pinned after calling this method (for example tooltip integration). The problem is that if we depend on some kind of property, or even sync event, there might be a chance that `isVisible` is switched to `false` before exiting `.pin()` method:

```ts
...
balloon.pin({
});

ballon.isVisible // is false
...
```

Reply:
I'm not a huge fan of this RAF either but I understand your concerns and saw errors caused by the synchronous `checkVisibility()` call here.

I iterated on this solution and came up with another one that makes `BPV#attachTo()` bulletproof to invisible `HTMLElements` and does not require the initial `checkVisible()` because this branch of code is never reached (if `attachTo()` failed). The `_pinWhenIsVisibleCallback` upon `change:isVisible` also is not activated. On top of it, I successfully removed a resize observer from the `TooltipManager`, too. 

Tests passed and I couldn't break the balloon system by testing manually. There are few tests missing for full CC, though.

You can see my approach here https://github.com/ckeditor/ckeditor5/compare/ck/16739...ck/16739-v2?expand=1.

Let me know what you think about it.

---

On a side note, if not for backward compatibility and unavoidable breaking changes, I'd warn if the pinning/attaching target is an invisible element because this whole logic of silently allowing divisible targets is a hack. You cannot position an element relative to another one that misses a position itself.